pkg/controller/endpointslicemirroring/reconciler.go

```diff
@@ -0,0 +1,405 @@
+		numExistingEndpoints += len(existingSlice.Endpoints)
+	}
+
+	// Build data structures for desired state.
+	desiredPortsByKey := map[addrTypePortMapKey][]discovery.EndpointPort{}
+	desiredEndpointsByKey := map[addrTypePortMapKey]endpointSet{}
+	numDesiredEndpoints := 0
+
+	for _, subset := range endpoints.Subsets {
+		endpointPorts := epPortsToEpsPorts(subset.Ports)
+		v4Hash := newAddrTypePortMapKey(endpointPorts, discovery.AddressTypeIPv4)
+		v6Hash := newAddrTypePortMapKey(endpointPorts, discovery.AddressTypeIPv6)
+		if _, ok := desiredEndpointsByKey[v4Hash]; !ok {
+			desiredEndpointsByKey[v4Hash] = endpointSet{}
+		}
+		if _, ok := desiredEndpointsByKey[v6Hash]; !ok {
+			desiredEndpointsByKey[v6Hash] = endpointSet{}
+		}
+
+		desiredPortsByKey[v4Hash] = endpointPorts
+		desiredPortsByKey[v6Hash] = endpointPorts
+		totalAddresses := 0
+
+		for _, address := range subset.Addresses {
+			totalAddresses++
+			if totalAddresses > maxEndpointsPerSubset {
+				break
+			}
+			endpoint := addressToEndpoint(address, true)
+			addrType := getAddressType(address.IP)
+			if addrType == nil {
+				klog.Warningf("Address in %s/%s Endpoints is not a valid IP, it will not be mirrored to an EndpointSlice: %s", endpoints.Namespace, endpoints.Name, address.IP)
+			} else {
+				epHash := v4Hash
+				if *addrType == discovery.AddressTypeIPv6 {
+					epHash = v6Hash
+				}
+				desiredEndpointsByKey[epHash].Insert(endpoint)
+				numDesiredEndpoints++
+			}
+		}
+
+		for _, address := range subset.NotReadyAddresses {
+			totalAddresses++
+			if totalAddresses > maxEndpointsPerSubset {
+				break
+			}
+			endpoint := addressToEndpoint(address, false)
+			addrType := getAddressType(address.IP)
+			if addrType == nil {
+				klog.Warningf("Address in %s/%s Endpoints is not a valid IP, it will not be mirrored to an EndpointSlice: %s", endpoints.Namespace, endpoints.Name, address.IP)
+			} else {
+				epHash := v4Hash
+				if *addrType == discovery.AddressTypeIPv6 {
+					epHash = v6Hash
+				}
+				desiredEndpointsByKey[epHash].Insert(endpoint)
+				numDesiredEndpoints++
+			}
+		}
+	}
+
+	epMetrics := mirroringmetrics.NewEndpointPortCache()
+	totalAdded := 0
+	totalRemoved := 0
+
+	// Determine changes necessary for each group of slices by port map.
+	for portKey, desiredEndpoints := range desiredEndpointsByKey {
+		numEndpoints := len(desiredEndpoints)
+		pmSlicesToCreate, pmSlicesToUpdate, pmSlicesToDelete, added, removed := r.reconcileByPortMapping(
+			endpoints, existingSlicesByKey[portKey], desiredEndpoints, desiredPortsByKey[portKey], portKey.addressType())
+
+		totalAdded += added
+		totalRemoved += removed
+
+		epMetrics.Set(endpointutil.PortMapKey(portKey), mirroringmetrics.EfficiencyInfo{
+			Endpoints: numEndpoints,
+			Slices:    len(existingSlicesByKey[portKey]) + len(pmSlicesToCreate) - len(pmSlicesToDelete),
+		})
+
+		if len(pmSlicesToCreate) > 0 {
+			slicesToCreate = append(slicesToCreate, pmSlicesToCreate...)
```

Comment:
if the `if` required? append of empty slice is nop?

Reply:
Updated, thanks!